### PR TITLE
lab 2, part 1

### PR DESCRIPTION
**Pull request by RenchTG**

### CODE REVIEW

**Comment by RenchTG** [comment]
So the executable seems to be built and linked and stored somewhere in nix files. But the `result` directory is not actually created. Though at this point I think it's just me not fully understanding how easy_cmake is being implemented. Any help on this error would be much appreciated:
```
error: builder for '/nix/store/ayq3rway90jzfm39fgm84wdmcs61hyjj-hello_world_lib-0.1.0.drv' failed with exit code 2;
       last 25 log lines:
       >     CMAKE_INSTALL_INCLUDEDIR
       >     CMAKE_INSTALL_INFODIR
       >     CMAKE_INSTALL_LIBDIR
       >     CMAKE_INSTALL_LIBEXECDIR
       >     CMAKE_INSTALL_LOCALEDIR
       >     CMAKE_INSTALL_MANDIR
       >     CMAKE_INSTALL_OLDINCLUDEDIR
       >     CMAKE_INSTALL_SBINDIR
       >     CMAKE_POLICY_DEFAULT_CMP0025
       >
       >
       > -- Build files have been written to: /build/d939yccw8dh0rf5zg50yyq259qc6h2mm-source/build
       > cmake: enabled parallel building
       > cmake: enabled parallel installing
       > Running phase: buildPhase
       > build flags: -j16 SHELL=/nix/store/4bj2kxdm1462fzcc2i2s4dn33g2angcc-bash-5.2p32/bin/bash
       > [ 25%] Building CXX object CMakeFiles/hello_world_lib.dir/src/hello_world.cpp.o
       > [ 50%] Linking CXX shared library libhello_world_lib.so
       > [ 50%] Built target hello_world_lib
       > [ 75%] Building CXX object CMakeFiles/test.dir/src/main.cpp.o
       > [100%] Linking CXX executable test
       > [100%] Built target test
       > Running phase: installPhase
       > install flags: -j16 SHELL=/nix/store/4bj2kxdm1462fzcc2i2s4dn33g2angcc-bash-5.2p32/bin/bash install
       > make: *** No rule to make target 'install'.  Stop.
       For full logs, run 'nix log /nix/store/ayq3rway90jzfm39fgm84wdmcs61hyjj-hello_world_lib-0.1.0.drv'.
```
Every other solution I've seen so far uses overlays and I wanted to try without them. Is maybe that the issue?

File: PhilipDobranowski/lab2_part1/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.24)
+
+project(hello_world LANGUAGES CXX)
+
+add_library(hello_world_lib SHARED src/hello_world.cpp)
+target_include_directories(hello_world_lib PUBLIC include)
+
+add_executable(test src/main.cpp)
+target_link_libraries(test PUBLIC hello_world_lib)
```
Comment (SreeDan) [inline PhilipDobranowski/lab2_part1/CMakeLists.txt:9]: Nice job on this, you are so close! You are missing one step in your CMake description. You have successfully built the project, but you are forgetting one step. Hint: currently, your CMake description does not add your built package to the list of all your current downloaded packages. This is the reason why you are getting your error.

**Review by RCMast3r — CHANGES_REQUESTED** [review]

File: PhilipDobranowski/lab2_part1/CMakeLists.txt (hunk 1 above)
Comment (RenchTG) [reply PhilipDobranowski/lab2_part1/CMakeLists.txt:9]: Why is this now needed in this lab, but wasn't in lab 1? Does this allow Nix to access our library when it's currently only accessible to our binary?

Comment (RenchTG) [reply PhilipDobranowski/lab2_part1/CMakeLists.txt:9]: > Nice job on this, you are so close! You are missing one step in your CMake description. You have successfully built the project, but you are forgetting one step. Hint: currently, your CMake description does not add your built package to the list of all your current downloaded packages. This is the reason why you are getting your error.

Why is this now needed in this lab, but wasn't in lab 1? Does this allow Nix to access our library when it's currently only accessible to our binary?

File: PhilipDobranowski/lab2_part1/flake.nix [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+  description = "Hello World part 1 flake";
+
+  inputs = {
+    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
+    easy_cmake.url = "github:RCMast3r/easy_cmake";
```
[the remark this answers is not in the document]
Comment (RenchTG) [reply PhilipDobranowski/lab2_part1/flake.nix:6]: Sounds good, I removed it. Though I have a question. How do we know what cmake commands Nix is actually running? How does `nativeBuildInputs = [ cmake ];` execute both `cmake` and `make -j`. I see some examples online specify commands in a `buildPhase` and `installPhase` variable, so how does this work without it?

[the remark this answers is not in the document]
Comment (RenchTG) [reply PhilipDobranowski/lab2_part1/flake.nix:6]: Also, what version of cmake does the derivation use and where is it getting it from? Is it a part of nixpkgs that is getting input to the derivation?

**Comment by RenchTG** [comment]
> Nice job on this, you are so close! You are missing one step in your CMake description. You have successfully built the project, but you are forgetting one step. Hint: currently, your CMake description does not add your built package to the list of all your current downloaded packages. This is the reason why you are getting your error.

Ok so it looks like what was actually causing this issue is that Nix was running `make install` under the hood and I had no idea and that was unaccounted for in my CMakeLists.txt. I guess `nativeBuildInputs = [ cmake ];` on its own runs `cmake`, `make`, and `make install` all automatically. Would be nice to know where to find that this is actually happening. But I opted to just explicitly define the commands so I know what's happening.

**Comment by RenchTG** [comment]
Then for this to start working I also had to change the `add_library` command in my CMakeLists to be `STATIC` instead of `SHARED` to avoid the error: `RPATH of binary /nix/store/c7vi75f5x3kvc7jdqlpb4hriarmw8acv-hello_world_lib-0.1.0/bin/test contains a forbidden reference to /build/`. Not sure what is happening here, also it looks like a .a file is being generated instead of a .so??

**Comment by RenchTG** [comment]
Ignore all previous questions, I got it figured out. my changes just need to be reviewed now

**Review by RCMast3r — APPROVED** [review]

**Pull request merged**